sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -0,0 +1,465 @@
+    }
+
+    public <X> Read<X> withRowMapper(RowMapper<X> rowMapper) {
+      return new Read<X>(options, rowMapper);
+    }
+
+    public Read<T> withUsername(String username) {
+      return new Read<T>(options.withUsername(username), rowMapper);
+    }
+
+    public Read<T> withPassword(String password) {
+      return new Read<T>(options.withPassword(password), rowMapper);
+    }
+
+    private final JdbcOptions options;
+    private final RowMapper<T> rowMapper;
+
+    private Read(JdbcOptions options, RowMapper<T> rowMapper) {
+      this.options = options;
+      this.rowMapper = rowMapper;
+    }
+
+    @Override
+    public PCollection<T> apply(PBegin input) {
+      PCollection<T> output = input.apply(Create.of(options))
+          .apply(ParDo.of(new ReadFn<>(rowMapper)));
+
+      return output;
+    }
+
+    @Override
+    public void validate(PBegin input) {
+      Preconditions.checkNotNull(rowMapper, "rowMapper");
+      options.validate();
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+
+      options.populateDisplayData(builder);
+    }
+
+    @VisibleForTesting
+    static class JdbcOptions implements Serializable {
+
+      private final DataSource dataSource;
+      private final String query;
+      @Nullable
+      private final String username;
+      @Nullable
+      private final String password;
+
+      private JdbcOptions(DataSource dataSource, String query,
+                          @Nullable String username,
+                          @Nullable String password) {
+        this.dataSource = dataSource;
+        this.query = query;
+        this.username = username;
+        this.password = password;
+      }
+
+      public JdbcOptions withDataSource(DataSource dataSource) {
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public JdbcOptions withQuery(String query) {
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public JdbcOptions withRowMapper(RowMapper rowMapper) {
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public JdbcOptions withUsername(String username) {
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public JdbcOptions withPassword(String password) {
+        return new JdbcOptions(dataSource, query, username, password);
+      }
+
+      public void validate() {
+        Preconditions.checkNotNull(dataSource, "dataSource");
+        Preconditions.checkNotNull(query, "query");
+      }
+
+      public void populateDisplayData(DisplayData.Builder builder) {
+        builder.add(DisplayData.item("dataSource", dataSource.getClass().getName()));
+        builder.add(DisplayData.item("query", query));
+        builder.addIfNotNull(DisplayData.item("username", username));
+      }
+
+      public DataSource getDataSource() {
```

Comment:
These getters seem unnecessary - it's a private class essentially. The rest of JdbcIO has access to its variables.